### PR TITLE
escape: Replace space characters in --data-urlencode value with +

### PR DESCRIPTION
**Pull request by bbakersmith**
Addresses  KNOWN_BUGS 4.5 Improve --data-urlencode space encoding and #3229

### CODE REVIEW

File: include/curl/curl.h [hunk 1]
```diff
@@ -2375,6 +2375,9 @@ CURL_EXTERN char *curl_easy_escape(CURL *handle,
 CURL_EXTERN char *curl_escape(const char *string,
                               int length);
 
+CURL_EXTERN char *curl_easy_escape_form(CURL *handle,
+                                        const char *string,
+                                        int length);
```
Comment (bagder) [inline include/curl/curl.h]: I'm not convinced we want a new public API for this, and if we do, we need docs and more tests. But yes, I realize that if it isn't a new public API we need to figure how to get this done in another way.

File: lib/escape.c [hunk 2]
```diff
@@ -64,6 +64,16 @@ bool Curl_isunreserved(unsigned char in)
   return FALSE;
 }
 
+/* spaces in form data are replaced with + rather than standard URL encoding
+ * %20, so consider spaces unreserved so they can be handled specially later on
+ */
+bool form_isunreserved(unsigned char in)
```
Comment (bagder) [inline lib/escape.c]: this should be static and its prototype should not be in the header file

File: lib/escape.c [hunk 3]
```diff
@@ -132,6 +144,28 @@ char *curl_easy_escape(struct Curl_easy *data, const char *string,
...
+
+/* escape form data per RFC1866 */
+char *curl_easy_escape_form(struct Curl_easy *data, const char *string,
+                            int inlength)
+{
+  char *enc = curl_easy_escape_flexible(data, string, inlength,
+                                        &form_isunreserved);
+
+  /* replace space with + */
+  for(int enc_index = 0; enc_index < strlen(enc); enc_index++) {
```
Comment (bagder) [inline lib/escape.c]: variable declarations within the `for()` is not C89 compliant.

I think the strlen() call should rather be done once outside of the loop than for every lap.

File: lib/escape.c [hunk 4]
```diff
@@ -132,6 +144,28 @@ char *curl_easy_escape(struct Curl_easy *data, const char *string,
...
+    if(*(enc + enc_index) == ' ')
```
Comment (bagder) [inline lib/escape.c]: I prefer the style `enc[enc_index]` to index a byte.

File: lib/escape.c [hunk 5]
```diff
@@ -76,8 +86,10 @@ char *curl_unescape(const char *string, int length)
   return curl_easy_unescape(NULL, string, length, NULL);
 }
 
-char *curl_easy_escape(struct Curl_easy *data, const char *string,
-                       int inlength)
+/* allow for custom definition of unreserved characters */
+char *curl_easy_escape_flexible(struct Curl_easy *data, const char *string,
```
Comment (bagder) [inline lib/escape.c]: this is a private function, right? It should be static and not use the `curl_` prefix.

**Comment by bbakersmith** [comment]
Thanks for the review! I made the changes you pointed out, very helpful. I was also hesitant to add to the public API for this, but it ultimately seemed the most straightforward approach. Please let me know if you have any thoughts on ways we could avoid this.

**Comment by bagder** [comment]
One simple but ineffective way would be to just encode as usual and then process the encoded output string a second time and then replace all '%20' instances with '+'. It would then leave the library untouched.

**Comment by bbakersmith** [comment]
Ah of course, thank you. I will open a new PR using that approach.

**Pull request closed without merging**